# Week 9 Assignment
## By: Natan Bienstock
## Tidying and Transforming Data

In this project, I tidy a flights dataset taken from a csv file based on data from Numbersense. <br>
The dataset contains arrival delay information for two airlines to five destinations. <br>
I clean up the data, and transform the data from "wide" to "long" format. <br>
Using this transformed and tidied data, I am able to answer some questions about and perform some analysis on the data. <br>
Finally, I transform the "long" data into a "wider" format that I believe would be the best way to visually represent the data. 

In [1]:
import pandas as pd

## Problem 2: Tidying and Transforming the Flights Data
First I read the flights data from the flights.csv file, that I created and uploaded to my GitHub account (Problem 1), into a Pandas DataFrame. <br>
I then output the flights DataFrame to the console. <br>

In [2]:
# read flights.csv file into pandas data frame
flights = pd.read_csv('https://raw.githubusercontent.com/ngb0330/DAV-5400/master/Week%209/flights.csv')
# return flights data frame
flights

Unnamed: 0 Unnamed: 1  Los Angeles Phoenix  San Diego  San Francisco Seattle
0     ALASKA    on time        497.0     221      212.0          503.0   1,841
1        NaN    delayed         62.0      12       20.0          102.0     305
2        NaN        NaN          NaN     NaN        NaN            NaN     NaN
3    AM WEST    on time        694.0   4,840      383.0          320.0     201
4        NaN    delayed        117.0     415       65.0          129.0      61

From the output of the flights DataFrame, I can see that I will need to do some work to tidy the data.<br>
First of all, there is an entire row of data that is missing from the DataFrame. <br>
This is due to there being a blank line in the csv file <br>
The first thing I thus do to clean the data, is removed that blank line, using the dropna function. <br>

In [3]:
# drop all rows from flights dataframe where every value
# in the row is null. 
flights.dropna(how = 'all', inplace = True)
# return flights dataframe. 
flights

Unnamed: 0 Unnamed: 1  Los Angeles Phoenix  San Diego  San Francisco Seattle
0     ALASKA    on time        497.0     221      212.0          503.0   1,841
1        NaN    delayed         62.0      12       20.0          102.0     305
3    AM WEST    on time        694.0   4,840      383.0          320.0     201
4        NaN    delayed        117.0     415       65.0          129.0      61

After removing the row that had only null values, I can see that there are still some null values in my dataframe. <br>
These null values come in the first column of data, where the airline names are listed. <br>
This is because in the original csv file, these values are not specified, as it is implied that they refer to the previous airline specified. <br>
I used the fillna function (using forward fill method), to replace these null values with the previous airline mentioned, as was implied in the original dataset. 

In [4]:
# fill all null values using forward fill
flights.fillna(method = 'ffill', inplace = True)
# return flights dataframe.
flights

Unnamed: 0 Unnamed: 1  Los Angeles Phoenix  San Diego  San Francisco Seattle
0     ALASKA    on time        497.0     221      212.0          503.0   1,841
1     ALASKA    delayed         62.0      12       20.0          102.0     305
3    AM WEST    on time        694.0   4,840      383.0          320.0     201
4    AM WEST    delayed        117.0     415       65.0          129.0      61

After filling all null values, I now decided to rename the first two columns. <br>
Since in the original dataset, the first two columns did not have names, when importing the data, the columns were assigned the names Unnamed: 0 and Unnamed: 1. <br>
I rename those two columns to be Airline and Status respectively using the rename function. 

In [5]:
# rename first two columns of flights dataframe to
# Airline and Status
flights.rename(columns = {'Unnamed: 0': 'Airline', 'Unnamed: 1': 'Status'}, inplace = True)
# return flights dataframe.
flights

Airline   Status  Los Angeles Phoenix  San Diego  San Francisco Seattle
0   ALASKA  on time        497.0     221      212.0          503.0   1,841
1   ALASKA  delayed         62.0      12       20.0          102.0     305
3  AM WEST  on time        694.0   4,840      383.0          320.0     201
4  AM WEST  delayed        117.0     415       65.0          129.0      61

#### Transforming DataFrame from "Wide" to "Long" Format
After ensuring that all null values were taken care of and all columns were properly named, I needed to convert the data from a wide format to a long format. <br>
This needs to be done as the data is untidy, as it has values for column names (some of the columns should really be values of a specific column), and there is more than one observation per row. <br>
To tidy the data, I used pandas melt function on the dataframe, to convert the columns Los Angeles, Phoenix, San Diego, San Francisco, and Seattle into one column called Destination. <br>
I keep the counts under each column, and put them into a new column called Flights. <br>
This data now contains one observation per row, and the column names are all variables and not values. 
The first 5 rows of the new dataframe, stored in the flights_long variable, are shown below. <br>

In [6]:
# Combine the Los Angeles, Phoenix, San Diego, San Francisco, and Seattle 
# columns into a new Destination column and create new column called
# Flights to hold the counts from the melted columns.
flights_long = pd.melt(flights, ['Airline', 'Status'], 
        value_vars = ['Los Angeles', 'Phoenix', 'San Diego', 'San Francisco', 'Seattle' ], 
       var_name = 'Destination', value_name = 'Flights')
# return first 5 rows of melted data. 
flights_long.head()

Airline   Status  Destination Flights
0   ALASKA  on time  Los Angeles     497
1   ALASKA  delayed  Los Angeles      62
2  AM WEST  on time  Los Angeles     694
3  AM WEST  delayed  Los Angeles     117
4   ALASKA  on time      Phoenix     221

After converting the data to "long" format, I still needed to do one more transformation in order to analyze the data. <br>
As shown below, all columns were of type object (or string) because some of the original flight numbers had comma values in them (example: 1,841). <br>
So I need to convert the Flights values to be numeric types in order to perform math on them. <br>
To do so, I first convert the entire Flights column to strings. <br>
Then I replace all "," characters with nothing (so "1,841" becomes "1841). <br>
Then I convert the Flights column to float type. <br>
I return the data types of the columns again to show that Flights column is now float type. <br>

In [7]:
# return data type of columns in flight_long dataframe.
flights_long.dtypes

Airline        object
Status         object
Destination    object
Flights        object
dtype: object

In [8]:
# change data type of Flights column to str.
flights_long.Flights = flights_long.Flights.astype(str)
# replace all ',' characters with nothing ''.
flights_long.Flights = flights_long.Flights.str.replace(',', '')
# convert Flights column to float 
flights_long.Flights = flights_long.Flights.astype(float)
# return data types of columns in flights_long dataframe.
flights_long.dtypes

Airline         object
Status          object
Destination     object
Flights        float64
dtype: object

## Problem 3 Analyzing the Transformed Flights Data
### Which Airline Had the Best Performance For Each City?
To answer this question, I created a copy of the flights dataframe to manipulate to perform my analysis. <br>
I then made a variable flights_total, using pandas' group by and sum functions to get the total number of flights per airlines per destination. <br>
I then added the Total Flights column that I just created per city into my flights_copy dataframe, by merging the flights_copy and flights_total dataframes on the Airline and Destination columns. <Br>
My flights_copy dataframe now has a column called Total Flights, in addition to the number of flights that were on time or delayed. <br>
I then found the percentage of flights that were each status (either on time or delayed) by dividing the Flights column by the Total Flights column. <br>
I created a dataframe dest_max that holds the maximum percentage of on time flights for each destination. <br>
I then use a for loop to loop through the dest_max dataframe, and print out each row where the maximum on time percentage was found in the flights_copy dataframe. <br>
This analysis showed that Alaska had the highest on time percentage for each individual city in the dataset. <br>

In [9]:
# Create a copy of the flights_long dataframe.
flights_copy = flights_long.copy()
# Return the first five rows of flights_copy.
flights_copy.head()

Airline   Status  Destination  Flights
0   ALASKA  on time  Los Angeles    497.0
1   ALASKA  delayed  Los Angeles     62.0
2  AM WEST  on time  Los Angeles    694.0
3  AM WEST  delayed  Los Angeles    117.0
4   ALASKA  on time      Phoenix    221.0

In [10]:
# Group the flights_copy data by Airline and Destination
# Sum the Flights column for each Airline and Destination.
flights_total = flights_copy.groupby(['Airline', 'Destination']).sum().reset_index()
# Return the first five rows of flights_total data.
flights_total.head()

Airline    Destination  Flights
0  ALASKA    Los Angeles    559.0
1  ALASKA        Phoenix    233.0
2  ALASKA      San Diego    232.0
3  ALASKA  San Francisco    605.0
4  ALASKA        Seattle   2146.0

In [11]:
#Rename the Flights column in flights_total to Total Flights.
flights_total.rename(columns = {'Flights': 'Total Flights'}, inplace = True)
# Return first five rows of flights_total data.
flights_total.head()

Airline    Destination  Total Flights
0  ALASKA    Los Angeles          559.0
1  ALASKA        Phoenix          233.0
2  ALASKA      San Diego          232.0
3  ALASKA  San Francisco          605.0
4  ALASKA        Seattle         2146.0

In [12]:
# Merge the flights_copy and flights_total dataframes
# on the Airline and Destination columns
# This adds in the Total Flights column to the flights_copy dataframe.
flights_merged = pd.merge(flights_copy,flights_total, on = ['Airline', 'Destination'])
# Return first five rows of flights_merged data.
flights_merged.head()

Airline   Status  Destination  Flights  Total Flights
0   ALASKA  on time  Los Angeles    497.0          559.0
1   ALASKA  delayed  Los Angeles     62.0          559.0
2  AM WEST  on time  Los Angeles    694.0          811.0
3  AM WEST  delayed  Los Angeles    117.0          811.0
4   ALASKA  on time      Phoenix    221.0          233.0

In [13]:
# Create column Status Percentage by dividing Flights by Total Flights
flights_merged['Status Percentage'] = flights_merged['Flights']/flights_merged['Total Flights']
# Return first rows of flights_merged data.
flights_merged.head()

Airline   Status  Destination  Flights  Total Flights  Status Percentage
0   ALASKA  on time  Los Angeles    497.0          559.0           0.889088
1   ALASKA  delayed  Los Angeles     62.0          559.0           0.110912
2  AM WEST  on time  Los Angeles    694.0          811.0           0.855734
3  AM WEST  delayed  Los Angeles    117.0          811.0           0.144266
4   ALASKA  on time      Phoenix    221.0          233.0           0.948498

In [14]:
# Group the flights_merged dataframe by Destination, and take the maximum Status Percentage
# for each Destination where its status is on time.
dest_max = flights_merged[flights_merged['Status'] == 'on time'].groupby(['Destination'])['Status Percentage'].max()
# Return the dest_max data
dest_max

Destination
Los Angeles      0.889088
Phoenix          0.948498
San Diego        0.913793
San Francisco    0.831405
Seattle          0.857875
Name: Status Percentage, dtype: float64

In [15]:
# Loop through the dest_max dataframe
# and Print out the flights_merged row that has Status Percentage 
# equal to the dest_max percentage for each row. 
for i in range(dest_max.shape[0]):
    print(flights_merged[flights_merged['Status Percentage'] == dest_max[i]])

  Airline   Status  Destination  Flights  Total Flights  Status Percentage
0  ALASKA  on time  Los Angeles    497.0          559.0           0.889088
  Airline   Status Destination  Flights  Total Flights  Status Percentage
4  ALASKA  on time     Phoenix    221.0          233.0           0.948498
  Airline   Status Destination  Flights  Total Flights  Status Percentage
8  ALASKA  on time   San Diego    212.0          232.0           0.913793
   Airline   Status    Destination  Flights  Total Flights  Status Percentage
12  ALASKA  on time  San Francisco    503.0          605.0           0.831405
   Airline   Status Destination  Flights  Total Flights  Status Percentage
16  ALASKA  on time     Seattle   1841.0         2146.0           0.857875


### Which Airline Had the Best Overall Performance
To answer this question, I created a subset of the data called on_time_flights that contained only flights that were on time. <br>
I then grouped this data by airline and found the sum of the Flights and Total Flights columns using pandas' groupby and sum functions. <br>
I then added a column to this data (stored in the total_flights_on_time variable) called On Time Percentage found by dividing the Flights column by Total Flights. <br>
I then sorted the data in descending order by the On Time Percentage and returned the top result. <br>
The analysis revealed that overall AM West had the highest percentage of On Time Flights with 89.11% of on time flights. <br>

In [16]:
# Create subset of flights_merged dataframe that 
# holds only on time flights.
on_time_flights = flights_merged[flights_merged['Status'] == 'on time']
# Return the on_time_flights dataframe.
on_time_flights

Airline   Status    Destination  Flights  Total Flights  Status Percentage
0    ALASKA  on time    Los Angeles    497.0          559.0           0.889088
2   AM WEST  on time    Los Angeles    694.0          811.0           0.855734
4    ALASKA  on time        Phoenix    221.0          233.0           0.948498
6   AM WEST  on time        Phoenix   4840.0         5255.0           0.921028
8    ALASKA  on time      San Diego    212.0          232.0           0.913793
10  AM WEST  on time      San Diego    383.0          448.0           0.854911
12   ALASKA  on time  San Francisco    503.0          605.0           0.831405
14  AM WEST  on time  San Francisco    320.0          449.0           0.712695
16   ALASKA  on time        Seattle   1841.0         2146.0           0.857875
18  AM WEST  on time        Seattle    201.0          262.0           0.767176

In [17]:
# Group the on_time_flights data by Airline, and sum the Flights
# and Total Flights columns.
total_flights_on_time = on_time_flights[['Airline','Flights','Total Flights']].groupby('Airline').sum()
# Return the total_flights_on_time dataframe.
total_flights_on_time

Flights  Total Flights
Airline                        
ALASKA    3274.0         3775.0
AM WEST   6438.0         7225.0

In [18]:
# Create column On Time Percentage by dividing Flights by Total Flights
total_flights_on_time['On Time Percentage'] = total_flights_on_time['Flights']/total_flights_on_time['Total Flights']
# Return the total_flights_on_time dataframe.
total_flights_on_time

Flights  Total Flights  On Time Percentage
Airline                                            
ALASKA    3274.0         3775.0            0.867285
AM WEST   6438.0         7225.0            0.891073

In [19]:
# Sort the data by On Time Percentage descending
# and display the top result.
total_flights_on_time.sort_values(by = 'On Time Percentage',ascending = False).head(1)

Flights  Total Flights  On Time Percentage
Airline                                            
AM WEST   6438.0         7225.0            0.891073

## Problem 4: Visually Presenting the Data
In order to best present the data, I decided to reorganize the data from my tidy data frame, to something a little less tidy. <br>
I made the data a little wider, and reorganized it so that the data is grouped first by Destination and then by Airline. <br>
The other columns show the number of on time flights and the number of delayed flights. <br>
I decided to organize the data this way because I believe that somebody looking at this data would want to first look at the destination level (because they are planning to go somewhere), and then at the airline level. <br>
In other words, you would want to look at a destination (ex. Los Angeles) and then which airline to take to get there (ex. Alaska). <br>
To choose an airline, you would want to see how many on time flights that airline had to that destination, which is why the columns are on time and then delayed. <br>
I believe that this is the best way to visually present the data, as it is how people would search for information about airlines to their chosen destination. <br>

In [20]:
# Set the index of the flights_long dataframe to be Airline and Destination, 
# and store this transformation in flights_long2.
flights_long2 = flights_long.set_index(['Airline','Destination'])
# Pivot the data so that Status is made into a column (unmelt the Status column).
pivoted = flights_long2.pivot_table(index = ['Airline', 'Destination'], columns = 'Status', values = 'Flights')
# Return the pivoted data.
pivoted

Status                 delayed  on time
Airline Destination                    
ALASKA  Los Angeles       62.0    497.0
        Phoenix           12.0    221.0
        San Diego         20.0    212.0
        San Francisco    102.0    503.0
        Seattle          305.0   1841.0
AM WEST Los Angeles      117.0    694.0
        Phoenix          415.0   4840.0
        San Diego         65.0    383.0
        San Francisco    129.0    320.0
        Seattle           61.0    201.0

In [21]:
# Swap the levels of the dataframe so it is grouped by destination and then airline.
pivoted = pivoted.swaplevel('Destination', 'Airline').sort_values('Destination')
# return the pivoted dataframe.
pivoted

Status                 delayed  on time
Destination   Airline                  
Los Angeles   ALASKA      62.0    497.0
              AM WEST    117.0    694.0
Phoenix       ALASKA      12.0    221.0
              AM WEST    415.0   4840.0
San Diego     ALASKA      20.0    212.0
              AM WEST     65.0    383.0
San Francisco ALASKA     102.0    503.0
              AM WEST    129.0    320.0
Seattle       ALASKA     305.0   1841.0
              AM WEST     61.0    201.0

#### How to Visually Present the Data

In [22]:
# swap the order of the on time and delayed columns in the pivoted dataframe.
pivoted = pivoted[['on time', 'delayed']]
# return the pivoted dataframe.
pivoted

Status                 on time  delayed
Destination   Airline                  
Los Angeles   ALASKA     497.0     62.0
              AM WEST    694.0    117.0
Phoenix       ALASKA     221.0     12.0
              AM WEST   4840.0    415.0
San Diego     ALASKA     212.0     20.0
              AM WEST    383.0     65.0
San Francisco ALASKA     503.0    102.0
              AM WEST    320.0    129.0
Seattle       ALASKA    1841.0    305.0
              AM WEST    201.0     61.0